### PR TITLE
OpenMP: Don't use chunk size for static scheduling on GNU compiler

### PR DESCRIPTION
**Pull request by rgayatri23**
The PR fixed the regression discussed in #5600

### CODE REVIEW

File: cmake/kokkos_enable_options.cmake [hunk 1]
```diff
@@ -65,6 +65,9 @@ KOKKOS_ENABLE_OPTION(HIP_MULTIPLE_KERNEL_INSTANTIATIONS OFF "Whether multiple ke
 KOKKOS_ENABLE_OPTION(IMPL_DESUL_ATOMICS   ON  "Whether to use desul based atomics - option only during beta")
 KOKKOS_ENABLE_OPTION(DESUL_ATOMICS_EXTERNAL OFF "Whether to use an external desul installation")
 
+# Switched on by default
+KOKKOS_ENABLE_OPTION(IMPL_OPENMP_THREADPOOL_ONLY ON "To use native OpenMP implementation post 3.6 version")
```
Comment (crtrott) [inline cmake/kokkos_enable_options.cmake:69]: Change comment to: "Use legacy OpenMP thread-pool based loop implementation."

**Review by crtrott — CHANGES_REQUESTED** [review]

File: cmake/kokkos_enable_options.cmake (hunk 1 above)
Comment (cz4rs) [reply cmake/kokkos_enable_options.cmake:69]: Should the legacy option be `ON` by default?
I think we could  use refactored (native OpenMP) version by default and only introduce this to deal with the regression for the affected systems.

**Review by dalg24 — DISMISSED** [review]
Please include this in `OpenMP::print_configuration`

File: core/src/OpenMP/Kokkos_OpenMP_Instance.cpp [hunk 2]
```diff
@@ -427,6 +427,13 @@ void OpenMP::print_configuration(std::ostream &os, bool /*verbose*/) const {
   os << "no\n";
 #endif
 
+  os << " KOKKOS_ENABLE_IMPL_OPENMP_THREADPOOL_ONLY: ";
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Instance.cpp]:
```suggestion
  os << "OpenMP Options:\n";
  os << "  KOKKOS_ENABLE_IMPL_OPENMP_THREADPOOL_ONLY: ";
```

File: core/src/OpenMP/Kokkos_OpenMP_Instance.cpp [hunk 3]
```diff
@@ -384,68 +392,4 @@ bool OpenMPInternal::verify_is_initialized(const char *const label) const {
...
-      name, Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{1}, []() {});
-}
-
-namespace Impl {
-
-int g_openmp_space_factory_initialized =
-    initialize_space_factory<OpenMP>("050_OpenMP");
-
-}  // namespace Impl
-
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Instance.cpp:450]: Rebase gone wrong

**Comment by cz4rs** [comment]
Segfault in `KokkosCore_PerfTestExec` (`default_exec.overlap_range_policy`) is caused by https://github.com/kokkos/kokkos/commit/e4cce6e44.
This was not discovered in CI because all of our builds were running native OpenMP implementation.

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 4]
```diff
@@ -101,8 +106,13 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
   std::enable_if_t<!std::is_same<typename Policy::schedule_type::type,
                                  Kokkos::Dynamic>::value>
   execute_parallel() const {
+#ifdef KOKKOS_COMPILER_GNU
```
Comment (cz4rs) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:109]: We already skip the chunk size for NVHPC (not supported) in lines 38 - 42. You could stick with `KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE` macro by checking for GNU there.

Comment (dalg24) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:109]: It looks like this patch is only wanting to drop the chunk size for static scheduling.  The `KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE` is also used (12L above) with a dynamic schedule.

Comment (cz4rs) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:109]: True, my bad :grin:

File: Makefile.kokkos [hunk 5]
```diff
@@ -1,7 +1,7 @@
 # Default settings common options.
 
-KOKKOS_VERSION_MAJOR = 3
-KOKKOS_VERSION_MINOR = 7
+KOKKOS_VERSION_MAJOR = 4
+KOKKOS_VERSION_MINOR = 0
```
Comment (dalg24) [inline Makefile.kokkos:4]: undo

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp (hunk 4 above)
Comment (rgayatri23) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:109]: Yeah the dynamic schedule does not show any regression so I did not do anything for it.

**Comment by rgayatri23** [comment]
Retest this please.

**Comment by cz4rs** [comment]
> Retest this please.

@rgayatri23 The history is still messed up. This could probably be just one commit with the changes from `Kokkos_OpenMP_Parallel.hpp`?

**Review by cz4rs — APPROVED** [review]
Looks good to me.

**Review by crtrott — APPROVED** [review]

**Pull request merged**